airbyte-cdk/bulk/core/load/src/integrationTest/kotlin/io/airbyte/cdk/mock_integration_test/MockDestinationBackend.kt

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.mock_integration_test
+
+import io.airbyte.cdk.test.util.DestinationDataDumper
+import io.airbyte.cdk.test.util.OutputRecord
+
+object MockDestinationBackend {
+    private val lock = Object()
+    private val files: MutableMap<String, MutableList<OutputRecord>> = mutableMapOf()
+
+    fun insert(filename: String, vararg records: OutputRecord) {
+        synchronized(lock) { getFile(filename).addAll(records) }
```

Comment:
I think we should use the kotlin Mutex::withLock style here. I'm not sure these play well with coroutines?

Reply:
I have no idea if this is correct, but: https://blog.danlew.net/2020/01/28/coroutines-and-java-synchronization-dont-mix/

Reply:
blegh. I do know that synchronized doesn't play well with virtual threads, so I wouldn't be surprised that coroutines also don't like it. Will fix.

Reply:
actually hm - I think this is fine? b/c none of these methods will yield at any point, so there's no risk of bad things happening?

(I tried naively switching to mutex.lock + make everything a suspend fun, which caused the test to block :P which is probably fixable, but would maybe require every test to be written as `runTest {withContext(IO) { process.run() ...  }}`)

Reply:
So just naively replacing every `synchronized(lock) { ... }` with `myMutex.withLock { ... }` blocked?

That's problematic.

If the only calls to the back end are via `insert` in `processBatch`, (b/c the dumper runs at the end, right?). What's blocking what?